test/integration/multicluster/install_test.go

```diff
@@ -165,63 +173,83 @@ func TestMulticlusterResourcesPostInstall(t *testing.T) {
 		{Namespace: "linkerd-multicluster", Name: "linkerd-gateway"},
 	}
 
-	for _, ctx := range contexts {
-		TestHelper.SwitchContext(ctx)
-		testutil.TestResourcesPostInstall(TestHelper.GetMulticlusterNamespace(), multiclusterSvcs, testutil.MulticlusterDeployReplicas, TestHelper, t)
-	}
+	TestHelper.SwitchContext(contexts[testutil.TargetContextKey])
+	testutil.TestResourcesPostInstall(TestHelper.GetMulticlusterNamespace(), multiclusterSvcs, testutil.MulticlusterDeployReplicas, TestHelper, t)
 }
 
 func TestLinkClusters(t *testing.T) {
-	linkName := "target"
-	// Get gateway IP from target cluster
-	lbCmd := []string{
-		"get", "svc",
-		"-n", "kube-system", "traefik",
-		"-o", "go-template={{ (index .status.loadBalancer.ingress 0).ip }}",
-	}
-	lbIP, err := TestHelper.KubectlWithContext("", contexts[testutil.TargetContextKey], lbCmd...)
+	linkCmds := make(map[string][]string)
+
+	// For each context construct the link command arguments
+	//  Source cluster will be linked against a target cluster without a gateway
+	//  Target cluster will be linked against a source cluster with a gateway and
+	//  with headless services
+	for k, ctx := range contexts {
+		// Get gateway IP from target cluster
+		lbCmd := []string{
+			"get", "node",
```

Comment:
can you explain this change from using the traefik address to using the node address?

Reply:
ofc! It looks like a scary change and it's a bit subtle. We rely on `traefik` to get the external IP we can use to reach the API server address. There's a problem here though, since our build system sets up [the source cluster without a load balancer](https://github.com/linkerd/linkerd2/blob/main/justfile#L180) ([we don't add it in CI either](https://github.com/linkerd/linkerd2/blob/main/.github/workflows/integration.yml#L380-L391)). 

We can special case this and use a lb (and traefik) in CI but locally the tests are going to be hard to run without being aware that our tests rely on traefik. There's a better way to do this that doesn't make us depend on a cluster with a load balancer config.

[k3s](https://docs.k3s.io/architecture) runs its API server on the control plane node. k3d wraps everything in a docker container. The API server address will actually correspond to the docker container's IP address, and since our two clusters run in a shared network, they can directly address each other. We can test this out by looking at the `kubernetes` service's EndpointSlice object:

```yaml
addressType: IPv4
apiVersion: discovery.k8s.io/v1
endpoints:
- addresses:
  - 192.168.160.2
  conditions:
    ready: true
kind: EndpointSlice
metadata:
  creationTimestamp: "2023-10-10T16:40:46Z"
  generation: 1
  labels:
    kubernetes.io/service-name: kubernetes
  name: kubernetes
  namespace: default
  resourceVersion: "73"
  uid: 76cb8809-1e9d-4910-9a04-1bc017e7fe42
ports:
- name: https
  port: 6443
  protocol: TCP
```

The address is the same as the node:

```
:; k get no -l node-role.kubernetes.io/control-plane=true -o go-template="{{ (index (index .items 0).status.addresses 0).address }}"
192.168.160.2⏎ 
```

For "regular" kubectl interactions, k3d tends to have a "serverlb" docker container that opens up a port locally and forwards connections to the associated control plane container.

Whenever the container is reachable it can be used to talk to the API server on port 6443.

Reply:
ah, interesting.  is this behavior specific to k3d?  e.g. does this make it so these tests can't be run on a kind cluster or on a cloud cluster?  I don't want to bog down this change but it sounds like we may have some debt here to make sure these tests don't overly rely on specific kubernetes implementations.

Reply:
`traefik` is definitely k3d/k3s specific (it's their default installation). Cloud clusters would definitely not work without some tweaking since their API servers are generally hosted and exposed to the internet (there's no need to override it for a link). I'm unsure about kind. It would depend on where it runs its API server; if it's on the node, then this change will make it work interchangeably.

As it stands, there is debt here, yes.